wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+				$this->c         = $this->input[ $this->position ] ?? null;
+				$this->n         = $this->input[ $this->position + 1 ] ?? null;
+				$this->type      = self::AT_TEXT_SUFFIX;
+			} else {
+				$this->consume();
+				$this->type = self::AT_SIGN_SYMBOL;
+			}
+		} elseif ( '?' === $la ) {
+			$this->consume();
+			$this->type = self::PARAM_MARKER;
+		} elseif ( '\\' === $la ) {
+			if ( 'N' === $la2 ) {
+				$this->consume(); // Consume the '\'.
+				$this->consume(); // Consume the 'N'.
+				$this->type = self::NULL2_SYMBOL;
+			} else {
+				$this->consume();
+				$this->type = self::INVALID_INPUT;
+			}
+		} elseif ( '#' === $la ) {
+			$this->line_comment();
+		} elseif ( $this->is_whitespace( $la ) ) {
+			while ( $this->is_whitespace( $this->c ) ) {
+				$this->consume();
+			}
+			$this->channel = self::CHANNEL_HIDDEN;
+		} elseif ( '0' === $la && ( 'x' === $la2 || 'b' === $la2 ) ) {
+			$this->number();
+		} elseif ( ( 'x' === $la || 'X' === $la || 'b' === $la || 'B' === $la ) && "'" === $la2 ) {
+			$this->number();
+		} elseif ( ( 'n' === $la || 'N' === $la ) && "'" === $la2 ) {
+			$prefix = $la;
+			$this->consume(); // n/N
+			$this->single_quoted_text();
+			if ( self::SINGLE_QUOTED_TEXT === $this->type ) {
+				$this->text = $prefix . $this->text;
+				$this->type = self::NCHAR_TEXT;
+			}
+		} elseif ( preg_match( '/\G' . self::PATTERN_UNQUOTED_IDENTIFIER . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$p               = $this->position - 1;
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+
+			// When preceded by a dot, it is always an identifier.
+			if ( $p >= 0 && '.' === $this->input[ $p ] ) {
+				$this->type = self::IDENTIFIER;
+			} elseif ( '_' === $la && isset( self::UNDERSCORE_CHARSETS[ strtolower( $this->text ) ] ) ) {
+				$this->type = self::UNDERSCORE_CHARSET;
+			} else {
+				$this->identifier_or_keyword();
+			}
+		} elseif ( null === $la ) {
+			$this->match_eof();
+			$this->token_instance = new MySQL_Token( self::EOF, '<EOF>' );
+			return false;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+
+		$this->token_instance = null === $this->type ? null : new MySQL_Token( $this->type, $this->text, $this->channel );
+		return true;
+	}
+
+	protected function la( int $i ) {
+		return $this->input[ $this->position + $i - 1 ] ?? null;
+	}
+
+	protected function consume() {
+		$this->text     .= $this->c;
+		$this->position += 1;
+		$this->c         = $this->input[ $this->position ] ?? null;
+		$this->n         = $this->input[ $this->position + 1 ] ?? null;
+	}
+
+	protected function match_eof() {
+		if ( null === $this->c ) {
+			$this->consume();
+		} else {
+			throw new RuntimeException( 'Current character is not EOF.' );
+		}
+	}
+
+	protected function identifier_or_keyword() {
```

Comment:
Is this about classifying the current token? It would be useful to use a name distinguishing it from functions like `number()` that also advance the parser.